### PR TITLE
Performance improvements/optimizations (#1371)

### PR DESCRIPTION
**Pull request by alexristich**
Tweaks relating to #1371.
This PR includes the following:

1) Removal of `onResponseStarted` listener, as this behavior is handled in the `onBeforeSendHeaders` listener. Tested to verify that it is indeed not needed.
2) Migration of `updateCount` listener from an event listener to places in the code where we actually cancel requests. NB: This change currently prevents the green `0` badge counter from showing, as it will not be triggered if no requests take place. I can think of a couple workarounds to this (_e.g._ calling this when the "tab" is updated), though nothing is currently implemented.
3) Moving of non-blocking `onBeforeSendHeaders` listener in `heuristicblocking.js` into the same listener in `webrequest.js`.

I've done some performance benchmarking of this, but I'm finding it very challenging to find reliable results. I've been doing 5 page loads in a row for the current production version, then for this version, then comparing the results. However, when I rerun this same experiment I'm finding the results can be wildly different for both. I haven't found this version to be any slower than the production version in any cases, though I haven't been able to gather sufficiently concrete statistics as to what the speedup might be, if any. Any tips here would be greatly appreciated.

The ratio is about 3:1 / 4:1 in terms of Chrome extensions overhead to Privacy Badger application code. This means for each 100ms of CPU time consumed by Privacy Badger, it costs about 300-400ms of CPU time in internal calls.

Lastly, with the `onBeforeRequest`, `onBeforeSendHeaders` and `onHeadersReceived` listeners all being required, there is a limit to how much we can speed things up. I see room for some more modest improvements in Privacy Badger's, but it is unlikely we'll get rid of the huge plateau with the current blocking model (_i.e._ up to 3 blocking event listeners for each request).

cpu plateau

### CODE REVIEW

**Comment by ghostwords** [comment]
Nice! I'm a fan of this PR already as it removes a bunch of code. I wonder if we'll be able to notice these speedups better in Firefox. How long does it take to visit X websites (with Selenium maybe) without Privacy Badger, with released Privacy Badger, with development Privacy Badger?

File: src/js/webrequest.js [hunk 1]
```diff
@@ -103,6 +103,7 @@ function onBeforeRequest(details){
       };
       chrome.tabs.sendMessage(tab_id, msg);
 
+      badger.updateCount(details);
```
Comment (ghostwords) [inline src/js/webrequest.js]: Should probably make updating count async by moving it behind a `setTimeout`.

File: src/js/webrequest.js [hunk 2]
```diff
@@ -103,6 +103,7 @@ function onBeforeRequest(details){
       };
       chrome.tabs.sendMessage(tab_id, msg);
 
+      badger.updateCount(details.tabId);
```
Comment (ghostwords) [inline src/js/webrequest.js]: Should probably make updating count async by moving it behind a `setTimeout`.

**Comment by alexristich** [comment]
One idea to re-introduce the `0` count for the badge is to call `updateCount` in the `main_frame` case.

**Comment by alexristich** [comment]
@ghostwords I'm having trouble here figuring out how to add the `updateCount` call in the `main_frame` case. 

When I add `badger.updateCount(tabId);` right after `forgetTab` [here](https://github.com/EFForg/privacybadger/blob/master/src/js/webrequest.js#L54), it creates a weird case where the green background with the `0` pops up for the badge, but quickly disappears. This is reproducible on any page with no registered trackers, _e.g._ https://www.eff.org/about.

I went through this many times with the debugger, but am having a lot of difficulty tracking down what might be causing the disappearance of the badge text in this case.

**Comment by ghostwords** [comment]
Could you revert the code style change? We should try to keep the change set focused (and the code style situation is in flux, [I wouldd rather not use Java style for non-objects](https://github.com/EFForg/privacybadger/pull/1403#discussion_r119754903)).

File: src/js/webrequest.js [hunk 3]
```diff
@@ -536,7 +536,7 @@ function checkAction(tabId, url, quiet, frameId){
   if (action && ! quiet) {
     badger.logTrackerOnTab(tabId, requestHost, action);
     if (_.contains(constants.BLOCKED_ACTIONS, action)) {
-      badger.updateCount(tabId);
+      setTimeout(badger.updateCount(tabId), 0);
```
Comment (ghostwords) [inline src/js/webrequest.js]: `updateCount` is immediately executed, returning a string, which is then passed into `setTimeout`. You need to do something like the following instead:
```javascript
setTimeout(function () {
  badger.updateCount(tab_id);
}, 0);
```

File: src/js/constants.js [hunk 4]
```diff
@@ -29,6 +29,13 @@ var exports = {
   USER_BLOCK: "user_block",
   USER_COOKIE_BLOCK: "user_cookieblock",
 
+  BLOCKED_ACTIONS: [
```
Comment (ghostwords) [inline src/js/constants.js]: Could you make it into an object made up of existing constants? Something like:
```javascript
constants.BLOCKED = {
  USER_BLOCK: true,
};
```
Checking then a `BLOCKED.hasOwnProperty()` call, avoiding array scanning.

File: src/js/background.js [hunk 5]
```diff
@@ -508,14 +508,13 @@ Badger.prototype = {
 
   /**
    * Checks conditions for updating page action badge and call updateBadge
-   * @param {Object} details details object from onBeforeRequest event
+   * @param {Object} tabId The tab ID whose badge we want to update.
    */
-  updateCount: function(details) {
-    if(!this.isPrivacyBadgerEnabled(webrequest.getHostForTab(details.tabId))){
+  updateCount: function(tabId) {
```
Comment (ghostwords) [inline src/js/background.js]: Let's go with `tab_id` as this is a string, not an object.

File: src/js/webrequest.js [hunk 6]
```diff
@@ -118,6 +118,12 @@ function onBeforeRequest(details){
  * @returns {*} modified headers
  */
 function onBeforeSendHeaders(details) {
+  if (badger) {
+    setTimeout(function() {
+      return badger.heuristicBlocking.heuristicBlockingAccounting(details);
```
Comment (ghostwords) [inline src/js/webrequest.js]: No need to use `return`, `heuristicBlockingAccounting` doesn't return anything and nobody will get the returned value out of a `setTimeout` anyway.

File: src/js/webrequest.js [hunk 7]
```diff
@@ -117,13 +117,19 @@ function onBeforeRequest(details){
  * @returns {*} modified headers
  */
 function onBeforeSendHeaders(details) {
-  let frame_id = details.frameId,
+  if (badger) {
+    setTimeout(function() {
+      return badger.heuristicBlocking.heuristicBlockingAccounting(details);
```
Comment (ghostwords) [inline src/js/webrequest.js]: No need to use return, heuristicBlockingAccounting doesn't return anything and nobody will get the returned value out of a setTimeout anyway.

**Comment by ghostwords** [comment]
I see what you mean about the badge being cleared soon after being set when set in `chrome.webRequest.onBeforeRequest` for `main_frame` types. Is this Chrome clearing it for us? The issue might be that `chrome.webRequest` is a bad fit for hooking into navigation events as perceived by the user, that's the job of [`chrome.webNavigation`](https://developer.chrome.com/extensions/webNavigation).
>In general, the webNavigation events are closely related to the navigation state that is displayed in the UI, while the webRequest events correspond to the state of the network stack which is generally opaque to the user.

**Comment by ghostwords** [comment]
We could back out `updateCount` tweaks for now; there is no need for them to go together with the other changes.

File: src/js/constants.js (hunk 4 above)
Comment (alexristich) [inline src/js/constants.js]: I tried at first to use existing constants, but the scoping wasn't working. If I declare the object like this:

```
  BLOCKED_ACTIONS: {
    BLOCK: true,
    COOKIEBLOCK: true,
    USER_BLOCK: true,
    USER_COOKIE_BLOCK: true
  };
```

then I do `Object.keys(constants.BLOCKED_ACTIONS)` it returns this:

```
["BLOCK", "COOKIEBLOCK", "USER_BLOCK", "USER_COOKIE_BLOCK"]
```

An object makes sense though; shall I swap it to an object with the keys as the current elements in the array?

File: src/js/webrequest.js (hunk 3 above)
Comment (alexristich) [inline src/js/webrequest.js]: Thanks! I kept looking at this and had a sense something was wrong. I should've just brushed up on the documentation. :)

**Comment by alexristich** [comment]
Good point on the refactoring. I've reverted this commit, and will leave renaming/refactoring for another PR once the style has been nailed down.

Also good intuition on the badge. Found a pretty straightforward workaround using the `chrome.tabs.onUpdated` API method; will push now.

File: src/js/constants.js (hunk 4 above)
Comment (ghostwords) [inline src/js/constants.js]: Sure, but you can do:
```javascript
exports.BLOCKED = {
  USER_BLOCK: true,
  ...
};
```

Comment (ghostwords) [inline src/js/constants.js]: Oops, sorry, my previous suggestion is wrong, "USER_BLOCK" is just a string then. You can do either
```javascript
let BLOCKED = {};
BLOCKED[exports.COOKIEBLOCK] = true;
BLOCKED[exports.BLOCK] = true;
...
exports.BLOCKED = BLOCKED;
```
or (using ES6 computed property names),
```javascript
exports.BLOCKED = {
  [exports.COOKIEBLOCK]: true,
  [exports.BLOCK]: true,
  ...
};
```

**Comment by ghostwords** [comment]
Lint checks are failing.

Tests against master ("continuous-integration/travis-ci/pr") break out linting into its job so you can still see how the other tests do. Tests in this branch ("continuous-integration/travis-ci/push") still use the old test setup.

File: src/js/constants.js [hunk 8]
```diff
@@ -29,6 +29,13 @@ var exports = {
   USER_BLOCK: "user_block",
   USER_COOKIE_BLOCK: "user_cookieblock",
 
+  BLOCKED_ACTIONS: {
+    "block": true,
```
Comment (ghostwords) [inline src/js/constants.js]: Please see https://github.com/EFForg/privacybadger/pull/1455#discussion_r124578177

File: src/js/webrequest.js [hunk 9]
```diff
@@ -706,6 +715,7 @@ function startListeners() {
   chrome.webRequest.onHeadersReceived.addListener(onHeadersReceived, {urls: ["<all_urls>"]}, ["responseHeaders", "blocking"]);
   chrome.tabs.onRemoved.addListener(onTabRemoved);
   chrome.tabs.onReplaced.addListener(onTabReplaced);
+  chrome.tabs.onUpdated.addListener(onTabUpdated);
```
Comment (ghostwords) [inline src/js/webrequest.js]: We already have a `chrome.tabs.onUpdated` listener in background.js ...

Comment (alexristich) [inline src/js/webrequest.js]: Yeah, we actually have some duplications still of the `chrome.tabs` listeners. To be fair, I'm really not helping the cause here though... :P 

- `chrome.tabs.onRemoved` appears in webrequest.js and incognito.js
- `chrome.tabs.onReplaced` appears in webrequest.js and background.js
- `chrome.tabs.onUpdated` appeared in background.js and incognito.js (and now in webrequest.js).

I wasn't too worried about these, as they're very infrequent in terms of the frequency at which they're called, but it shouldn't be too much trouble to merge them all. Should I go ahead and do this?

Comment (ghostwords) [inline src/js/webrequest.js]: Possibly, if that makes sense. Registering twice for the same thing seems like extra overhead although how much (a trivial amount?) is unclear. It might also make sense to split off the badge changes into their own PR.

Comment (alexristich) [inline src/js/webrequest.js]: Sure, I took a quick look and it doesn't seem to be too hard.

As far as pulling out the badge changes, it might be a bit tricky; looks like I'd have to pull out a number of non-linear commits. I do see your point though, as there seems to be enough work around the badge to warrant another PR, though I think it does still fit under the scope of performance improvements, since we're moving it to (a) only be called when new trackers are detected, and (b) be called in an async runtime.

I do have some thoughts about how the badge number is being counted. Currently, each time `updateBadge()` is called, we go through every third-party domain that has been loaded on the specified tab and find the ones with a stored action present in `BLOCKED_ACTIONS`. This is O(n^2) runtime, albeit with n being relatively small. I'm not sure whether improving this is a priority right now, though to improve it to linear runtime would require some changes to the `tabData` data structure that @cowlicks has been tweaking in another PR, so not entirely sure this PR would be the best place for that in any case.

File: src/js/background.js (hunk 5 above)
Comment (alexristich) [inline src/js/background.js]: Currently the first half(ish) of `webrequest.js` uses `tab_id` and the second half(ish) uses `tabId`. If it's cool with you I'll just leave this and we can have a separate PR that fixes the wonky parameter and variable styling.

File: src/js/constants.js (hunk 4 above)
Comment (alexristich) [inline src/js/constants.js]: Nice tip, and computed property names are pretty cool!

File: src/js/heuristicblocking.js [hunk 10]
```diff
@@ -499,44 +499,9 @@ function hasCookieTracking(details, origin) {
...
-    var hasSetCookie = false;
-    for(var i = 0; i < details.responseHeaders.length; i++) {
-      if(details.responseHeaders[i].name.toLowerCase() == "set-cookie") {
-        hasSetCookie = true;
-        break;
-      }
-    }
-    if(hasSetCookie) {
-      if (badger) {
-        return badger.heuristicBlocking.heuristicBlockingAccounting(details);
```
Comment (cowlicks) [inline src/js/heuristicblocking.js]: Here `heuristicBlockingAccounting` is removed in two places -- when we send headers, and receive a response. But it is only added back in an `onBeforeSendHeaders` listener. So I think this may loose some accounting.

Comment (alexristich) [inline src/js/heuristicblocking.js]: Not to worry :) This PR is starting to feel a bit messy, but `onResponseStarted` turned out to be redundant. It was calling `heuristicBlockingAccounting()` both before the request was sent and after a response was received. I tested and confirmed that removing this listener didn't affect the tracking accounting. This refers to a comment above and the discussion we had in last week's dev meeting.

**Comment by cowlicks** [comment]
I'm skeptical that there is much of a performance improvement in removing the listeners from `heuristicblocking.js` and moving the same code into a listener is `webrequests.js`. It seems like we are doing the same amount of work, but we just have fewer listeners.

The place I think we are doing the most unnecessary work is in the `onBeforeRequest` listener in `background.js`. We may be able to completely remove this. The work I did on `tabData` is pretty stale, so feel free to disregard it.

**Comment by alexristich** [comment]
Recall that the performance bottleneck is primarily from (a) blocking listeners and (b) overhead from internal Chrome code required for each event dispatched. Turns out the efficiency of the application logic in PB isn't the culprit, so improvements here are nice but perhaps secondary to the larger architectural changes.

I'd thought the same regarding `onBeforeRequest`, though I remember in last week's chat you mentioned this was important to keep as it allows us to cancel requests before the TLS handshake. Is there a way around this?

**Comment by alexristich** [comment]
Closing this PR; I've created two separate branches to split out the badge changes from the listener refactoring.
